### PR TITLE
Tree: cleanup cursor tests, and fix json cursor bug

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

This deduplicates some of the json cursor specific tests with the ones in the cursor test suite.

As part of this, testing of keys on specific types of nodes was reworked so it did not depend on builders from jsonableTrees so that they could be applied to specialized cursors like JsonCursor. This rework checked for keys on all the test tree nodes, and this exposed a bug which the old test suite missed: keys like "0" were showing up on strings in the json domain when using named field access but not in field iteration. A fix for this bug is also included.

### CODE REVIEW

File: packages/dds/tree/src/test/cursorTestSuite.ts [hunk 1]
```diff
@@ -727,24 +465,21 @@ export function testTreeCursor<TData, TCursor extends ITreeCursor>(config: {
...
 
-const unusedKey: FieldKey = symbolFromKey(brand("unusedKey"));
-
 /**
  * Test that cursor works as a cursor.
  * This does NOT test that the data the cursor exposes is correct,
  * it simply checks that the traversal APIs function, and that a few aspects of them conform with the spec.
  *
- * TODO: add testing for paths to this, or a second test in testCursors.
+ * TODO: add testing for paths to this, or a separate tests for it.
```
Comment (jenn-le) [inline packages/dds/tree/src/test/cursorTestSuite.ts:473]:
```suggestion
 * TODO: add testing for paths to this, or separate tests for it.
```

File: packages/dds/tree/src/test/cursorTestSuite.ts [hunk 2]
```diff
@@ -184,17 +179,63 @@ export interface SpecialCaseBuilder<TData> {
...
+ * @typeParam TCursor - Type of the cursor being tested.
+ */
+export function testSpecializedCursor<TData, TCursor extends ITreeCursor>(config: {
+    cursorName: string;
+    builders: SpecialCaseBuilder<TData>;
+    cursorFactory: (data: TData) => TCursor;
+    dataFromCursor: (cursor: ITreeCursor) => TData;
+    testData: readonly { name: string; data: TData; reference?: JsonableTree }[];
+}): Mocha.Suite {
+    return testTreeCursor(config);
```
Comment (jenn-le) [inline packages/dds/tree/src/test/cursorTestSuite.ts:193]: What's the purpose of wrapping this? The only difference I see is that dataFromCursor isn't optional here but I also don't see testTreeCursor being called elsewhere.

**Review by jenn-le — APPROVED** [review]

File: packages/dds/tree/src/test/cursorTestSuite.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/test/cursorTestSuite.ts:193]: There are two exported entry points to this test suite:
1. testGeneralPurposeTreeCursor: this is for cursors which can be constructed from any jsonable tree.
2. testSpecializedCursor: this is for cursors which can't handle all possible trees, for example JsonCursor and mt upcoming uniform chunk cursor.

**Review by jenn-le — APPROVED** [review]

File: packages/dds/tree/src/test/domains/json/jsonCursor.spec.ts [hunk 3]
```diff
@@ -48,252 +46,6 @@ const testCases: readonly [string, readonly JsonCompatible[]][] = [
     ],
 ];
 
-describe("JsonCursor", () => {
```
Comment (DLehenbauer) [inline packages/dds/tree/src/test/domains/json/jsonCursor.spec.ts:51]: This set of tests might be handy to point cursor users to as documentation, but again, your call.

**Review by DLehenbauer — APPROVED** [review]

**Pull request merged**